ext/atomic/index.md

```diff
@@ -0,0 +1,512 @@
+# Atomic Operations Extension
+
+This extension provides a means to perform multiple "operations" in a linear and
+atomic manner. Operations are a serialized form of the mutations allowed in the
+base JSON:API specification.
+
+Clients can send an array of operations in a single request. This extension
+guarantees that those operations will be processed in order and will either
+completely succeed or fail together.
+
+## <a href="#namespace" id="namespace" class="headerlink"></a> Namespace
+
+This extension uses the namespace `atomic`.
+
+> Note: JSON:API extensions can only introduce new document members using a
+> reserved namespace as a prefix.
+
+## <a href="#document-structure" id="document-structure" class="headerlink"></a> Document Structure
+
+A document that supports this extension **MAY** include any of the top-level
+members allowed by the base specification with the exception of `data` and
+`included`, which **MUST NOT** be included.
+
+In addition, such a document **MAY** include either of the following members,
+but not both:
+
+- `atomic:operations` - an array of one or more [operation
+  objects](#operation-objects).
+
+- `atomic:results` - an array of one or more [result objects](#result-objects).
```

Comment:
I've been writing a simple implementation of this and this flexibility is causing a bit of extra work. Would it make sense to heighten this requirement to:

> In addition, such a document **MUST** include one of the following members, but not both:

And above, changing:

> A document that supports this extension

to:

> A document using this extension

Reply:
For context, here's a snippet of my code:

```php
$supported_extensions = [
  'https://jsonapi.org/extensions/#atomic'
];
$unsupported_extensions = array_diff($extensions, $supported_extensions);
if (!empty($unsupported_extensions)) {
  throw new UnsupportedMediaTypeException(sprintf('The %s JSON:API media type extension is not supported', current($unsupported_extensions)));
}
if (in_array($supported_extensions[0], $extensions, TRUE)) {
  $request_document = Json::decode((string) $request->getContent());
  if (!isset($request_document['atomic:operations'])) {
    throw new BadRequestHttpException('Request documents using the %s JSON:API media type extension must include an `atomic:operations` top-level member.');
  }
}
```

It's that final `throw new BadRequestException()` that I'm concerned about. According to the language above, I think it's technically valid for a client to send a `Content-Type` header with the atomic extension URI, even when it's not including an `atomic:operations` member.

Reply:
Actually, I guess it's not that bad. All it means is that I have to do:

```php
$operations = isset($request_document['atomic:operations'])
  ? $request_document['atomic:operations']
  : [];
```